lightningd/notification.c

```diff
@@ -59,4 +62,40 @@ void notify_warning(struct lightningd *ld, struct log_entry *l)
 	json_object_end(n->stream); /* .warning */
 	jsonrpc_notification_end(n);
 	plugins_notify(ld->plugins, take(n));
-}
\ No newline at end of file
+}
+
+void notify_forward_event(struct lightningd *ld,
+			  const struct htlc_in *in,
+			  const struct htlc_out *out,
+			  enum forward_status state,
+			  enum onion_type failcode,
+			  struct timeabs *resolved_time)
+{
+	struct jsonrpc_notification *n =
+		jsonrpc_notification_start(NULL, notification_topics[3]);
+	/* Here is more neat to initial a forwarding structure than
+	 * to pass in a bunch of parameters directly*/
+	struct forwarding *cur = tal(tmpctx, struct forwarding);
+	cur->channel_in = *in->key.channel->scid;
+	cur->msat_in = in->msat;
+	if (out) {
+		cur->channel_out = *out->key.channel->scid;
+		cur->msat_out = out->msat;
+		assert(amount_msat_sub(&cur->fee, in->msat, out->msat));
+	} else {
+		cur->channel_out.u64 = 0;
+		cur->msat_out = AMOUNT_MSAT(0);
+		cur->fee = AMOUNT_MSAT(0);
+	}
+	cur->payment_hash = tal(cur, struct sha256_double);
+	cur->payment_hash->sha = in->payment_hash;
```

Comment:
@ZmnSCPxj  My thoughts about `struct sha256` and struct `sha256_double` are mainly reflected here.
The same payment hash data, the different structures we used. 
Before storing forward payment information into DB, we used `struct sha256`. And after storing, we used `struct sha256_double`, which is involved in `struct forwarding`:
https://github.com/ElementsProject/lightning/blob/2945b25b57db37dca31c10da34a4b1ca33d697c3/wallet/wallet.h#L189-L198
For another example, we used `struct sha256` for `payment_hash` in `struct wallet_payment`:
https://github.com/ElementsProject/lightning/blob/2945b25b57db37dca31c10da34a4b1ca33d697c3/wallet/wallet.h#L241-L262

Like [BOLT#11](https://github.com/lightningnetwork/lightning-rfc/blob/90faaeda89e34cb5257317981e793b47d6182f6d/11-payment-encoding.md#tagged-fields) said,

> set `payment_hash` to the `SHA2` 256-bit hash of the `payment_preimage` that will be given in return for payment.

Should we change the structure of `payment_hash` in `struct forwarding` to `struct sha256`?

Reply:
It does look quite wrong... yes, we should probably fix this in a new PR.